### DISCUSSION TITLE
[Keycloak 25.0.4] HTTP_RELATIVE_PATH after adding on the hostname ???

### DISCUSSION THREAD

**Discussion by boddedo** [keycloak/keycloak#32246] [Archive]
### Description

Hello !

I would like to know what KC_HTTP_RELATIVE_PATH is used for after adding the new "hostname" ?

also I added HTTP instead of HTTPS in the hostname and I have this error:
```
Key material not provided to setup HTTPS. Please configure your keys/certificates or start the server in development mode.
```
I don't remember to have seen this error before.

I use NGINX as a reverse proxy.

here is my config :
```
/usr/bin/podman container run --name keycloak
    --log-driver k8s-file
    --env KEYCLOAK_ADMIN=mail@example.org
    --env KEYCLOAK_ADMIN_PASSWORD=<ADMINPASSWORD>
    --env KC_DB=mariadb
    --env KC_DB_URL_HOST=IP.ADD.RES.S
    --env KC_DB_URL_PORT=3306
    --env KC_DB_URL_PROPERTIES=?autoReconnect=true&useSSL=false&allowPublicKeyRetrieval=true
    --env KC_DB_URL_DATABASE=keycloak
    --env KC_DB_USERNAME=keycloak
    --env KC_DB_PASSWORD=<DBPASSWORD>
    --env KC_FEATURES=recovery-codes,token-exchange,update-email,web-authn,persistent-user-sessions
    --env KC_HEALTH_ENABLED=true
    --env KC_HOSTNAME=http://subdomainone.example.org:8080/auth
    --env KC_HOSTNAME_ADMIN=http://subdomaintwo.example.org:8080/auth
    --env KC_HOSTNAME_STRICT=false
    --env KC_HOSTNAME_DEBUG=true
    --env KC_HTTP_RELATIVE_PATH=/auth
    --env KC_METRICS_ENABLED=true
    --env KC_PROXY_HEADERS=forwarded
    --env KC_PROXY_ADDRESS_FORWARDING=true
    --env KC_TRANSACTION_XA_ENABLED=true
    --env KC_LOG_LEVEL=info,org.keycloak.social.user_profile_dump:debug
    --env QUARKUS_TRANSACTION_MANAGER_ENABLE_RECOVERY=true
    --env QUARKUS_TRANSACTION_MANAGER_OBJECT_STORE_DIRECTORY=/tmp/ObjectStore
    --volume /keycloak/kc-data/keycloak.conf:/opt/keycloak/conf/keycloak.conf
    --volume /keycloak/kc-data/providers:/opt/keycloak/providers
    --volume /keycloak/kc-data/data:/opt/keycloak/data
    --volume /keycloak/kc-data/themes:/opt/keycloak/themes
    --log-opt path=/var/log/keycloak/keycloak.log
    --publish 8080:8080
    quay.io/keycloak/keycloak:25.0.4 start --log-level="info,org.keycloak.social.user_profile_dump:debug"
```

Regards.

**Comment by shawkins** [comment]
It helps to think of things as options that affect how the server serves (http/s options), and options to configure how a proxy exposes the server (hostname / proxy options). 

KC_HTTP_RELATIVE_PATH affects the server - it means that everything is served relative to that path.

> Key material not provided to setup HTTPS. Please configure your keys/certificates or start the server in development mode.

http-enabled=false must be explicitly set to use http endpoints. Without that option the server assumes you want to use TLS, thus the error.

**Reply by boddedo** [reply]
Okay thank you for your answer !

but if I add KC_HTTP_RELATIVE_PATH,
do I need to add that in the "KC_HOSTNAME" and "KC_HOSTNAME_ADMIN" too ?
because in the documentation it says you can add it but it doesn't work (I obtain 502 bad gateway, I'll check this error in multiple hour).

**Reply by shawkins** [reply]
> do I need to add that in the "KC_HOSTNAME" and "KC_HOSTNAME_ADMIN" too ?

Those depend on your proxy configuration, not KC_HTTP_RELATIVE_PATH. Is the admin hostname different than the regular hostname? Then you need to set hostname-admin. If they are the same, then no. Beyond that setting the hostname will depend upon what the proxy is changing - the hostname, port, path, etc. Please try to reread the docs on this https://www.keycloak.org/server/hostname

> because in the documentation it says you can add it but it doesn't work (I obtain 502 bad gateway, I'll check this error in multiple hour).

You need to look in the error log for the exact problem - is it forming the correct request (you should be able to hit that url directly with a browser or curl from the machine the proxy is on and have the initial request succeed), or something like a forbidden error that means your system isn't setup yet for an nginx proxy https://serverfault.com/questions/819423/reverse-proxy-nginx-bad-gateway
